x/exchange/orders_test.go

```diff
@@ -0,0 +1,2786 @@
+		{
+			name: "invalid denom in assets",
+			order: AskOrder{
+				MarketId: 1,
+				Seller:   sdk.AccAddress("another_address_____").String(),
+				Assets:   *coin(1, "x"),
+				Price:    *coin(42, "farnsworth"),
+			},
+			exp: []string{"invalid assets", "invalid denom: x"},
+		},
+		{
+			name: "zero-value assets",
+			order: AskOrder{
+				MarketId: 1,
+				Seller:   sdk.AccAddress("another_address_____").String(),
+				Assets:   sdk.Coin{},
+				Price:    *coin(42, "farnsworth"),
+			},
+			exp: []string{"invalid assets", "invalid denom: "},
+		},
+		{
+			name: "price denom in assets",
+			order: AskOrder{
+				MarketId: 1,
+				Seller:   sdk.AccAddress("another_address_____").String(),
+				Assets:   *coin(2, "farnsworth"),
+				Price:    *coin(42, "farnsworth"),
+			},
+			exp: []string{"invalid assets", "price denom farnsworth cannot also be the assets denom"},
+		},
+		{
+			name: "invalid seller settlement flat fee denom",
+			order: AskOrder{
+				MarketId:                1,
+				Seller:                  sdk.AccAddress("another_address_____").String(),
+				Assets:                  *coin(99, "bender"),
+				Price:                   *coin(42, "farnsworth"),
+				SellerSettlementFlatFee: coin(13, "x"),
+			},
+			exp: []string{"invalid seller settlement flat fee", "invalid denom: x"},
+		},
+		{
+			name: "zero seller settlement flat fee denom",
+			order: AskOrder{
+				MarketId:                1,
+				Seller:                  sdk.AccAddress("another_address_____").String(),
+				Assets:                  *coin(99, "bender"),
+				Price:                   *coin(42, "farnsworth"),
+				SellerSettlementFlatFee: coin(0, "nibbler"),
+			},
+			exp: []string{"invalid seller settlement flat fee", "nibbler amount cannot be zero"},
+		},
+		{
+			name: "negative seller settlement flat fee",
+			order: AskOrder{
+				MarketId:                1,
+				Seller:                  sdk.AccAddress("another_address_____").String(),
+				Assets:                  *coin(99, "bender"),
+				Price:                   *coin(42, "farnsworth"),
+				SellerSettlementFlatFee: coin(-3, "nibbler"),
+			},
+			exp: []string{"invalid seller settlement flat fee", "negative coin amount: -3"},
+		},
+		{
+			name: "multiple problems",
+			order: AskOrder{
+				Price:                   *coin(0, ""),
+				SellerSettlementFlatFee: coin(0, ""),
+			},
+			exp: []string{
+				"invalid market id",
+				"invalid seller",
+				"invalid price",
+				"invalid assets",
+				"invalid seller settlement flat fee",
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			err := tc.order.Validate()
+			assertions.AssertErrorContents(t, err, tc.exp, "Validate() error")
+		})
+	}
+}
+
+func TestAskOrder_CopyChange(t *testing.T) {
+	coin := func(amount int64, denom string) sdk.Coin {
+		return sdk.Coin{Denom: denom, Amount: sdkmath.NewInt(amount)}
+	}
+	coinP := func(amount int64, denom string) *sdk.Coin {
+		rv := coin(amount, denom)
+		return &rv
+	}
```

Comment:
Same as above. Not entirely sure if needed since we do have NewInt64Coin.

Reply:
`NewInt64Coin` returns a concrete `Coin` struct, and you can't take the reference of a function result. E.g. you can't do `&sdk.NewInt64Coin(...)`.